packages/@react-stately/numberfield/src/useNumberFieldState.ts

```diff
@@ -11,159 +11,240 @@
+  let startingValue = useMemo(() => {
+    if (!isNaN(value)) {
+      return value;
+    } else if (!isNaN(defaultValue)) {
+      return defaultValue;
+    }
+    return NaN;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+  let lastValDispatched = useRef(startingValue);
+  let smartOnChange = useCallback((val) => {
+    if (!isNaN(val) || !isNaN(lastValDispatched.current)) {
+      onChange?.(val);
+    }
+    lastValDispatched.current = val;
+  }, [lastValDispatched, onChange]);
 
-  const numberParser = useNumberParser();
-  const textValueFormatter = useNumberFormatter(formatOptions);
-  const inputValueFormatter = useNumberFormatter();
 
-  const [numberValue, setNumberValue] = useControlledState<number>(value, defaultValue || 0, onChange);
-  let initialInputValue = inputValueFormatter.format(numberValue);
-  const [inputValue, setInputValue] = useState(isNaN(value) && isNaN(defaultValue) ? '' : initialInputValue);
-  const [isValid, setIsValid] = useState(isInputValueValid(numberValue, maxValue, minValue));
+  let [numberValue, setNumberValue] = useControlledState<number>(value, isNaN(defaultValue) ? NaN : defaultValue, smartOnChange);
 
-  const minusSign = useRef('-');
+  let textValue = inputValueFormatter.format(numberValue);
+  let [inputValue, setInputValue] = useState(() => isNaN(numberValue) ? '' : textValue);
 
+
+  // this updates the field only if the formatter or number has changed,
+  // this should only run after increment, decrement, blur
+  // or when the formatter changes, but not because of the overrides, which is why we run on formatOptions
+  // I doubt that the formatOptions will change while a user is typing a number, but that might cause a problem
   useEffect(() => {
-    // Get the minus sign of the current locale to filter the input value
-    // Automatically updates the minus sign when numberFormatter changes
-    minusSign.current = inputValueFormatter.formatToParts(-11).find(p => p.type === 'minusSign').value;
-  }, [inputValueFormatter]);
+    setInputValue(isNaN(numberValue) ? '' : textValue);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [formatOptions, setInputValue, numberValue, locale]);
+
+  let parse = (value: string): number => {
+    // Set to empty state if input value is empty
+    if (!value.length) {
+      return NaN;
+    }
+    let newValue = numberParser.parse(value);
+    // if it failed to parse, then reset input to formatted version of current number
+    if (isNaN(newValue)) {
+      return NaN;
+    }
+
+    return newValue;
+  };
 
-  let increment = () => {
+  // this removes any not allowed characters from the input value
+  let cleanInputValue = cleanString(inputValue, numberParser.symbols, locale, currentNumeralSystem);
+  // Number parser doesn't know about min/max, so we must remove it ourselves
+  if (minValue >= 0) {
+    cleanInputValue = cleanInputValue.replace(minusSign, '');
```

Comment:
only if the min value is >= 0, isn't that what we want? we can't type in a minus, so why should we be able to paste on in? what happens if they paste in 'abc123'? would you expect 'abc' to appear?

Reply:
No, I mean if you paste `-30` and the min value is >= 0, it should just do nothing. And IMO when pasting 'abc123' nothing should happen as well since it's not a valid number. Randomly extracting digits doesn't seem very useful to me... Seems like that would simplify things here too, e.g. not needing to expose this `clean` function.

Reply:
did you try this?